Initial implementation for fancy devicememcpybatch

mfbalin opened the pull request:
This is my initial go at enabling fancy iterators for source buffers for the devicememcpy::batched algorithm, aiming to address #674. I didn't make any attempt at changing the byte copy abstraction, thus I expect the code to behave exactly the same way as if input memory buffers were passed. However, this may not be optimal because if a user passes an expensive source iterator, it will be accessed more times than necessary.

I am wondering if this way of enabling fancy iterators is acceptable at all or should I try to completely specialize the implementation depending on the type of the items being copied. Also, the code is currently not working for my use case at all so would appreciate it if I broke something obvious and someone can point it out.

elstehle commented on cub/device/dispatch/dispatch_batch_memcpy.cuh:
```diff
@@ -215,8 +269,8 @@ __launch_bounds__(int(ChainedPolicyT::ActivePolicy::AgentSmallBufferPolicyT::BLO
 {
-  static_assert(std::is_pointer<cub::detail::value_t<InputBufferIt>>::value,
-                "The batched memcpy only supports copying of memory buffers");
+  // static_assert(std::is_pointer<cub::detail::value_t<InputBufferIt>>::value,
+  //               "The batched memcpy only supports copying of memory buffers");
```
Since we plan to distinguish `DeviceMemcpy::Batched` and `DeviceCopy::Batched`, I think we can move this assert to the interface method `DeviceMemcpy::Batched`, with the error message hinting at `DeviceCopy::Batched`. E.g.,
```
"DeviceMemcpy::Batched only supports copying of memory buffers. Please consider using DeviceCopy::Batched instead."
```

mfbalin replied:
Done.

elstehle reviewed (changes requested):
Thanks a lot for your contribution, @mfbalin! This already looks great!

We believe that the generic iterator use case should get its own interface called `DeviceCopy::Batched` living in `cub/device/device_copy.cuh`. The two can keep the common implementation with the specializations you have introduced. Only a new interface with slightly different semantics: `DeviceCopy::Batched`'s `buffer_sizes` will be the number of _items_ (instead of number of _bytes_). This is just a matter of documentation. The idea is that `DeviceCopy::Batched` works with arbitrary iterators to iterators and will become the batched version of `std::copy`, whereas `DeviceMemcpy::Batched` is the batched version `std::memcpy`. 

Could you please also add tests that make sure the new functionality works as expected? I think we should also have a test where each buffer is a fancy output iterator, like a `TransformOutputIterator`, if possible.

Again, thanks a lot for your contribution to this project!

mfbalin commented:
> We believe that the generic iterator use case should get its own interface called `DeviceCopy::Batched` living in `cub/device/device_copy.cuh`. The two can keep the common implementation with the specializations you have introduced. Only a new interface with slightly different semantics: `DeviceCopy::Batched`'s `buffer_sizes` will be the number of _items_ (instead of number of _bytes_). This is just a matter of documentation. The idea is that `DeviceCopy::Batched` works with arbitrary iterators to iterators and will become the batched version of `std::copy`, whereas `DeviceMemcpy::Batched` is the batched version `std::memcpy`.

That was the bug I was experiencing in my use case, thanks for pointing out that memcpy requires buffer sizes in bytes, not items. The code works now, I will see how much performance improvement I will get for my devicerunlength::decode use case.

Is it not acceptable at all for devicememcpy to accept fancy iterators? When the input iterators are trivial to evaluate such as my use case where buffer `i` is made up of `[i, i, i, i, i]` by using:

```
struct RepeatIndex {
  template <typename IdType>
  __host__ __device__ auto operator()(IdType i) {
    return thrust::make_constant_iterator(i);
  }
};

thrust::counting_iterator<int64_t> iota(0);
auto input_buffer = thrust::make_transform_iterator(iota, RepeatIndex{});
```

Multiple reads of the exact same location in the buffer will be free in this use case assuming that everything will be inlined, and it will be perfectly reasonable to use memcpy over copy from a performance standpoint. I would imagine for the same task in this scenario, the performance of copy can only be equal to or slower than memcpy due to the variable size of the items being copied. If the items were to get too large, copy will have to use a single thread for each item, unlike memcpy which only sees bytes, reducing the amount of available parallelism. Or am I wrong and would it always be better to use a copy over memcpy for all fancy input iterator use cases?

mfbalin commented:
What else would I need to modify when it comes to the assignment of work to threads when switching from byte logic to item logic to implement a DeviceCopy?

mfbalin commented on cub/agent/agent_batch_memcpy.cuh:
```diff
@@ -789,6 +791,60 @@ private:
+      for (ByteOffsetT ichar = warp_lane; ichar < num_bytes; ichar += CUB_PTX_WARP_THREADS)
+      {
+        out_ptr[ichar] = read_byte(in_ptr, ichar);
+      }
```
Done.

elstehle commented:
> That was the bug I was experiencing in my use case, thanks for pointing out that memcpy requires buffer sizes in bytes, not items. The code works now, I will see how much performance improvement I will get for my devicerunlength::decode use case.
> Is it not acceptable at all for devicememcpy to accept fancy iterators? 

I think this shouldn't be supported. Even if it _was_ possible by the current implementation and users started relying on the `memcpy` supporting fancy iterators, it would limit future implementations for a `memcpy`. Providing two distinct interfaces, `DeviceCopy::Batched` and `DeviceMemcpy::Batched` gives the library the flexibility to further specialize the implementations in the future.

> When the input iterators are trivial to evaluate such as my use case where buffer `i` is made up of `[i, i, i, i, i]` by using:
> 
> ```
> struct RepeatIndex {
>   template <typename IdType>
>   __host__ __device__ auto operator()(IdType i) {
>     return thrust::make_constant_iterator(i);
>   }
> };
> 
> thrust::counting_iterator<int64_t> iota(0);
> auto input_buffer = thrust::make_transform_iterator(iota, RepeatIndex{});
> ```
> Multiple reads of the exact same location in the buffer will be free in this use case assuming that everything will be inlined,

The same is the case for the `Copy::Batched`, right?

>[...] and it will be perfectly reasonable to use memcpy over copy from a performance standpoint. I would imagine for the same task in this scenario, the performance of copy can only be equal to or slower than memcpy due to the variable size of the items being copied. If the items were to get too large, copy will have to use a single thread for each item, unlike memcpy which only sees bytes, reducing the amount of available parallelism. Or am I wrong and would it always be better to use a copy over memcpy for all fancy input iterator use cases?

I think there's no blanket answer here: It's reasonable to assume that it will be faster to copy whole elements of, e.g., `4 B`, `8 B` than to copy the bytes of those elements individually. Only once element's get _very_ large, I expect the byte-wise copy to gain an edge again. 

> What else would I need to modify when it comes to the assignment of work to threads when switching from byte logic to item logic to implement a DeviceCopy? Or could I just match each byte to an item and whenever a byte is read and written in the code, I can read and write an item in the code?
> P.S. That is the approach I am currently taking in the code.

Thanks, that looks exactly right to me. That's also what I have had in mind. 👍

elstehle commented on cub/agent/agent_batch_memcpy.cuh:
```diff
@@ -789,6 +795,80 @@ private:
+                                             OffsetT offset,
+                                             AliasT value)
+  {
+    reinterpret_cast<const AliasT *>(buffer_dst)[offset] = value;
```
```suggestion
    reinterpret_cast<AliasT *>(buffer_dst)[offset] = value;
```

elstehle reviewed (commented):
Thanks a lot for the great work! I'm really looking forward to having this land in CUB 🚀 

This is first part of the review. Just minor comments. I'll try to conclude the rest of the review today.

mfbalin commented:
The code already seems to work. Tuning parameters will probably need to be modified to take the type size into account. I think I can have the next round of reviews. I will work continue to work on the test.

mfbalin commented on cub/device/device_copy.cuh:
```diff
@@ -0,0 +1,174 @@
+  * required allocation size is written to \p temp_storage_bytes and no work is done.
+  * @param temp_storage_bytes [in,out] Reference to size in bytes of \p d_temp_storage allocation
+  * @param input_it [in] Device-accessible iterator providing the pointers to the source
+  * memory buffers
```
I simply used ranges.

mfbalin commented on cub/device/device_copy.cuh:
```diff
@@ -0,0 +1,174 @@
+  * ranges can alias one another.
+  *
+  * @par Snippet
+  * The code snippet below illustrates usage of DeviceCopy::Batched for mutating strings within
```
Updated the example.

elstehle commented on test/test_device_batch_copy.cu:
```diff
@@ -0,0 +1,427 @@
+struct RepeatIndex
+{
+  template <typename IdType>
+  __host__ __device__ auto operator()(IdType i)
```
For C++11 support.
```suggestion
  __host__ __device__ auto operator()(IdType i) -> thrust::constant_iterator<IdType>
```

elstehle commented on test/test_device_batch_copy.cu:
```diff
@@ -0,0 +1,427 @@
+  //---------------------------------------------------------------------
+  // The most granular type being copied. Buffer's will be aligned and their size be an integer
+  // multiple of this type
+  using AtomicCopyT = int64_t;
```
Since this is even more generic than the `DeviceMemcpy::Batched` it would be great if we could test for different item types. Like, `char`, `int64_t`, and one non-built-in type. For the non-built-in type you could use a `thrust::zip_iterator` to give you `thrust::tuple`s - or a different custom type, whichever you prefer.

mfbalin replied:
The test code currently includes different types such as int64_t, tuple<int64_t, int32_t int16_t, char, char>, uint8_t (at the end for the big buffer test),. The current code uses fancy iterators for both input and output.

elstehle commented on test/test_device_batch_copy.cu:
```diff
@@ -0,0 +1,427 @@
+/**
+ * @brief
+ *
+ * @tparam AtomicT The most granular type being copied. All source and destination pointers will be
+ * aligned based on this type, the number of bytes being copied will be an integer multiple of this
+ * type's size
```
```suggestion
* @tparam AtomicT The type of the elements being copied.
```

elstehle commented on test/test_device_batch_copy.cu:
```diff
@@ -0,0 +1,427 @@
+  // Type used for indexing into the array of buffers
+  using BufferOffsetT = uint32_t;
+
+  // Type used for indexing into individual bytes of a buffer (large enough to cover the max buffer
```
Could you double check? I think this refers to elements in the tests?
```suggestion
  // Type used for indexing into individual elements of a buffer (large enough to cover the max buffer
```

elstehle commented on test/test_device_batch_copy.cu:
```diff
@@ -0,0 +1,427 @@
+  constexpr BufferOffsetT single_buffer = 1;
+
+  // Run tests with input buffer being consecutive and output buffers being consecutive
+  RunTest<AtomicCopyT, BufferOffsetT, BufferSize64T, ByteOffset64T>(single_buffer,
+                                                                    large_target_copy_size,
+                                                                    large_target_copy_size,
+                                                                    TestDataGen::CONSECUTIVE);
```
I think this has quite some heavy memory requirements with a 64-bit `AtomicCopyT`. Maybe we can lower the memory requirements when checking 64-bit indexes by using a 8-bit`AtomicCopyT`?

elstehle reviewed (commented):
Great work, @mfbalin! Thanks a lot for the contribution and the excellent effort! I have left a few comments, mostly just minor ones. 

Could you also please add a test that where an output range is a fancy output iterator. E.g., a `TransformOutputIterator` that doubles the values being written. It can be a simple test, as the test we have for the 64-bit offsets. This is mostly just to ensure we didn't miss anything in the specialization and to make sure things compile fine.

elstehle commented:
> The code already seems to work. Tuning parameters will probably need to be modified to take the type size into account. I think I can have the next round of reviews. I will continue to work on the test.

Excellent! 🚀 Exactly, we can do the tuning policy work in a follow-up PR, if needed. 

I've run the existing benchmarks on the `DeviceMemcpy::Batched` to make sure we didn't introduce any major regression. Results for `DeviceMemcpy` seem untouched, which is good news 🙂 👍 

I'll check if I can extend the existing benchmarks to cover `DeviceCopy::Batched`.

```
|  AtomicT  |  Buffer Order  |  Min. buffer size  |  Max. buffer size  |  Elements  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |       Diff |   %Diff |  Status  |
|-----------|----------------|--------------------|--------------------|------------|------------|-------------|------------|-------------|------------|---------|----------|
|    U8     |     Random     |         1          |         8          |    2^25    |   4.083 ms |       0.50% |   4.080 ms |       0.50% |  -2.813 us |  -0.07% |   PASS   |
|    U8     |     Random     |         1          |         64         |    2^25    | 762.786 us |       0.45% | 762.437 us |       0.47% |  -0.350 us |  -0.05% |   PASS   |
|    U8     |     Random     |         1          |        256         |    2^25    | 414.391 us |       0.71% | 414.536 us |       0.74% |   0.146 us |   0.04% |   PASS   |
|    U8     |     Random     |         1          |        1024        |    2^25    | 348.240 us |       0.70% | 347.977 us |       0.75% |  -0.263 us |  -0.08% |   PASS   |
|    U8     |     Random     |         1          |       65536        |    2^25    | 202.079 us |       0.90% | 202.182 us |       0.85% |   0.103 us |   0.05% |   PASS   |
|    U8     |     Random     |       65536        |       65536        |    2^25    | 108.510 us |       1.24% | 108.478 us |       1.13% |  -0.031 us |  -0.03% |   PASS   |
|    U8     |     Random     |         1          |         8          |    2^27    |  18.515 ms |       0.56% |  18.516 ms |       0.50% |   0.706 us |   0.00% |   PASS   |
|    U8     |     Random     |         1          |         64         |    2^27    |   3.593 ms |       0.16% |   3.593 ms |       0.17% |  -0.066 us |  -0.00% |   PASS   |
|    U8     |     Random     |         1          |        256         |    2^27    |   1.802 ms |       0.27% |   1.803 ms |       0.28% |   0.866 us |   0.05% |   PASS   |
|    U8     |     Random     |         1          |        1024        |    2^27    | 828.094 us |       0.71% | 828.037 us |       0.67% |  -0.057 us |  -0.01% |   PASS   |
|    U8     |     Random     |         1          |       65536        |    2^27    | 536.573 us |       2.02% | 537.580 us |       1.90% |   1.008 us |   0.19% |   PASS   |
|    U8     |     Random     |       65536        |       65536        |    2^27    | 405.002 us |       2.03% | 405.484 us |       1.96% |   0.482 us |   0.12% |   PASS   |
|    U8     |     Random     |         1          |         8          |    2^29    |  77.478 ms |       0.50% |  77.471 ms |       0.49% |  -6.533 us |  -0.01% |   PASS   |
|    U8     |     Random     |         1          |         64         |    2^29    |  14.953 ms |       0.06% |  14.952 ms |       0.07% |  -0.181 us |  -0.00% |   PASS   |
|    U8     |     Random     |         1          |        256         |    2^29    |   7.393 ms |       0.11% |   7.390 ms |       0.11% |  -2.259 us |  -0.03% |   PASS   |
|    U8     |     Random     |         1          |        1024        |    2^29    |   3.404 ms |       0.34% |   3.404 ms |       0.36% |   0.376 us |   0.01% |   PASS   |
|    U8     |     Random     |         1          |       65536        |    2^29    |   1.854 ms |       1.05% |   1.849 ms |       0.92% |  -4.857 us |  -0.26% |   PASS   |
|    U8     |     Random     |       65536        |       65536        |    2^29    |   1.628 ms |       1.32% |   1.622 ms |       1.17% |  -6.064 us |  -0.37% |   PASS   |
|    U8     |  Consecutive   |         1          |         8          |    2^25    | 499.848 us |       0.38% | 499.943 us |       0.42% |   0.095 us |   0.02% |   PASS   |
|    U8     |  Consecutive   |         1          |         64         |    2^25    | 260.975 us |       0.83% | 262.726 us |       0.87% |   1.751 us |   0.67% |   PASS   |
|    U8     |  Consecutive   |         1          |        256         |    2^25    | 299.358 us |       1.10% | 299.998 us |       1.18% |   0.640 us |   0.21% |   PASS   |
|    U8     |  Consecutive   |         1          |        1024        |    2^25    | 308.905 us |       0.74% | 307.136 us |       0.83% |  -1.769 us |  -0.57% |   PASS   |
|    U8     |  Consecutive   |         1          |       65536        |    2^25    | 205.411 us |       1.04% | 205.487 us |       1.05% |   0.077 us |   0.04% |   PASS   |
|    U8     |  Consecutive   |       65536        |       65536        |    2^25    | 109.102 us |       1.33% | 109.209 us |       1.28% |   0.107 us |   0.10% |   PASS   |
|    U8     |  Consecutive   |         1          |         8          |    2^27    |   1.908 ms |       0.18% |   1.908 ms |       0.16% |  -0.273 us |  -0.01% |   PASS   |
|    U8     |  Consecutive   |         1          |         64         |    2^27    | 900.934 us |       0.23% | 903.329 us |       0.20% |   2.395 us |   0.27% |   FAIL   |
|    U8     |  Consecutive   |         1          |        256         |    2^27    |   1.093 ms |       0.53% |   1.095 ms |       0.53% |   1.862 us |   0.17% |   PASS   |
|    U8     |  Consecutive   |         1          |        1024        |    2^27    | 609.776 us |       0.79% | 612.644 us |       0.82% |   2.868 us |   0.47% |   PASS   |
|    U8     |  Consecutive   |         1          |       65536        |    2^27    | 536.132 us |       1.81% | 536.039 us |       1.64% |  -0.093 us |  -0.02% |   PASS   |
|    U8     |  Consecutive   |       65536        |       65536        |    2^27    | 406.250 us |       1.90% | 404.997 us |       1.95% |  -1.253 us |  -0.31% |   PASS   |
|    U8     |  Consecutive   |         1          |         8          |    2^29    |   7.560 ms |       0.79% |   7.558 ms |       0.50% |  -1.513 us |  -0.02% |   PASS   |
|    U8     |  Consecutive   |         1          |         64         |    2^29    |   3.471 ms |       0.16% |   3.469 ms |       0.18% |  -2.203 us |  -0.06% |   PASS   |
|    U8     |  Consecutive   |         1          |        256         |    2^29    |   4.197 ms |       0.33% |   4.199 ms |       0.28% |   1.169 us |   0.03% |   PASS   |
|    U8     |  Consecutive   |         1          |        1024        |    2^29    |   2.492 ms |       0.42% |   2.493 ms |       0.43% |   1.299 us |   0.05% |   PASS   |
|    U8     |  Consecutive   |         1          |       65536        |    2^29    |   1.834 ms |       0.95% |   1.830 ms |       0.87% |  -4.394 us |  -0.24% |   PASS   |
|    U8     |  Consecutive   |       65536        |       65536        |    2^29    |   1.649 ms |       1.66% |   1.642 ms |       1.29% |  -6.948 us |  -0.42% |   PASS   |
|    U32    |     Random     |         1          |         8          |    2^25    |   2.903 ms |       0.13% |   2.903 ms |       0.11% |  -0.616 us |  -0.02% |   PASS   |
|    U32    |     Random     |         1          |         64         |    2^25    | 724.012 us |       0.51% | 724.579 us |       0.47% |   0.567 us |   0.08% |   PASS   |
|    U32    |     Random     |         1          |        256         |    2^25    | 392.841 us |       0.77% | 391.940 us |       0.70% |  -0.901 us |  -0.23% |   PASS   |
|    U32    |     Random     |         1          |        1024        |    2^25    | 326.118 us |       0.80% | 325.245 us |       0.75% |  -0.873 us |  -0.27% |   PASS   |
|    U32    |     Random     |         1          |       65536        |    2^25    | 201.195 us |       1.01% | 200.773 us |       0.96% |  -0.422 us |  -0.21% |   PASS   |
|    U32    |     Random     |       65536        |       65536        |    2^25    | 107.419 us |       1.04% | 107.276 us |       1.17% |  -0.143 us |  -0.13% |   PASS   |
|    U32    |     Random     |         1          |         8          |    2^27    |  13.525 ms |       0.90% |  13.514 ms |       0.08% | -11.290 us |  -0.08% |   FAIL   |
|    U32    |     Random     |         1          |         64         |    2^27    |   3.409 ms |       0.18% |   3.409 ms |       0.16% |  -0.049 us |  -0.00% |   PASS   |
|    U32    |     Random     |         1          |        256         |    2^27    |   1.705 ms |       0.28% |   1.704 ms |       0.24% |  -0.721 us |  -0.04% |   PASS   |
|    U32    |     Random     |         1          |        1024        |    2^27    | 788.582 us |       0.73% | 788.906 us |       0.72% |   0.323 us |   0.04% |   PASS   |
|    U32    |     Random     |         1          |       65536        |    2^27    | 530.925 us |       1.81% | 530.575 us |       1.62% |  -0.350 us |  -0.07% |   PASS   |
|    U32    |     Random     |       65536        |       65536        |    2^27    | 405.204 us |       2.01% | 405.649 us |       1.94% |   0.446 us |   0.11% |   PASS   |
|    U32    |     Random     |         1          |         8          |    2^29    |  57.028 ms |       0.50% |  56.960 ms |       0.07% | -68.506 us |  -0.12% |   FAIL   |
|    U32    |     Random     |         1          |         64         |    2^29    |  14.206 ms |       0.06% |  14.207 ms |       0.06% |   1.451 us |   0.01% |   PASS   |
|    U32    |     Random     |         1          |        256         |    2^29    |   6.985 ms |       0.14% |   6.987 ms |       0.13% |   1.678 us |   0.02% |   PASS   |
|    U32    |     Random     |         1          |        1024        |    2^29    |   3.236 ms |       0.32% |   3.234 ms |       0.34% |  -2.147 us |  -0.07% |   PASS   |
|    U32    |     Random     |         1          |       65536        |    2^29    |   1.803 ms |       0.87% |   1.800 ms |       0.81% |  -3.331 us |  -0.18% |   PASS   |
|    U32    |     Random     |       65536        |       65536        |    2^29    |   1.617 ms |       1.26% |   1.611 ms |       1.07% |  -5.733 us |  -0.35% |   PASS   |
|    U32    |  Consecutive   |         1          |         8          |    2^25    | 385.902 us |       0.49% | 385.875 us |       0.49% |  -0.027 us |  -0.01% |   PASS   |
|    U32    |  Consecutive   |         1          |         64         |    2^25    | 246.558 us |       0.88% | 246.857 us |       0.86% |   0.298 us |   0.12% |   PASS   |
|    U32    |  Consecutive   |         1          |        256         |    2^25    | 286.994 us |       0.98% | 286.815 us |       0.86% |  -0.179 us |  -0.06% |   PASS   |
|    U32    |  Consecutive   |         1          |        1024        |    2^25    | 296.304 us |       0.92% | 292.909 us |       0.82% |  -3.395 us |  -1.15% |   FAIL   |
|    U32    |  Consecutive   |         1          |       65536        |    2^25    | 200.824 us |       1.22% | 200.875 us |       1.22% |   0.052 us |   0.03% |   PASS   |
|    U32    |  Consecutive   |       65536        |       65536        |    2^25    | 109.920 us |       1.46% | 109.924 us |       1.38% |   0.004 us |   0.00% |   PASS   |
|    U32    |  Consecutive   |         1          |         8          |    2^27    |   1.461 ms |       0.19% |   1.462 ms |       0.17% |   0.516 us |   0.04% |   PASS   |
|    U32    |  Consecutive   |         1          |         64         |    2^27    | 838.718 us |       0.28% | 838.587 us |       0.27% |  -0.130 us |  -0.02% |   PASS   |
|    U32    |  Consecutive   |         1          |        256         |    2^27    |   1.052 ms |       0.58% |   1.052 ms |       0.54% |   0.065 us |   0.01% |   PASS   |
|    U32    |  Consecutive   |         1          |        1024        |    2^27    | 601.856 us |       0.71% | 603.372 us |       0.71% |   1.516 us |   0.25% |   PASS   |
|    U32    |  Consecutive   |         1          |       65536        |    2^27    | 527.037 us |       1.22% | 526.328 us |       1.10% |  -0.709 us |  -0.13% |   PASS   |
|    U32    |  Consecutive   |       65536        |       65536        |    2^27    | 405.829 us |       1.91% | 404.364 us |       1.87% |  -1.465 us |  -0.36% |   PASS   |
|    U32    |  Consecutive   |         1          |         8          |    2^29    |   5.775 ms |       0.74% |   5.771 ms |       0.08% |  -4.083 us |  -0.07% |   PASS   |
|    U32    |  Consecutive   |         1          |         64         |    2^29    |   3.222 ms |       0.18% |   3.214 ms |       0.23% |  -8.087 us |  -0.25% |   FAIL   |
|    U32    |  Consecutive   |         1          |        256         |    2^29    |   4.048 ms |       0.35% |   4.050 ms |       0.35% |   1.800 us |   0.04% |   PASS   |
|    U32    |  Consecutive   |         1          |        1024        |    2^29    |   2.442 ms |       0.44% |   2.443 ms |       0.45% |   0.524 us |   0.02% |   PASS   |
|    U32    |  Consecutive   |         1          |       65536        |    2^29    |   1.807 ms |       0.90% |   1.805 ms |       0.81% |  -2.626 us |  -0.15% |   PASS   |
|    U32    |  Consecutive   |       65536        |       65536        |    2^29    |   1.638 ms |       1.42% |   1.631 ms |       1.09% |  -7.334 us |  -0.45% |   PASS   |
```

mfbalin commented on cub/device/dispatch/dispatch_batch_memcpy.cuh:
```diff
@@ -316,6 +326,7 @@ template <typename InputBufferIt,
           typename BufferSizeIteratorT,
           typename BufferOffsetT,
           typename BlockOffsetT,
+          bool memcpy             = true,
```
When the copy is called for a type whose assignment operator has been overloaded, won't this create a problem? Would I need to check both that it is a pointer and it is trivially copyable?

mfbalin replied:
Also, even if we infer from the types, the buffer sizes will be incorrect for trivially copyable types if a user called the Copy instead of Memcpy. How can we know whether Copy or Memcpy was called?

elstehle replied:
>Also, even if we infer from the types, the buffer sizes will be incorrect for trivially copyable types if a user called the Copy instead of Memcpy.

Thanks for bringing this up. You are right. We should maintain the `memcpy` template parameter to be able to identify whether the user has called the `DeviceCopy` or `DeviceMemcpy` interface and, hence, to correctly infer whether `size` would refer to `number_of_bytes` or `number_of_items`.

Could you put the `memcpy` template parameter last, though. That's to make sure that, if previously somebody was using the `Dispatch` class by providing the explicit template arguments, the template parameter order is unchanged and just extended.

mfbalin commented on cub/device/dispatch/dispatch_batch_memcpy.cuh:
```diff
@@ -415,8 +495,12 @@ struct DispatchBatchMemcpy : SelectedPolicy
+    using BlevBufferSrcsOutT =
+      typename ::cuda::std::conditional<memcpy, void *, cub::detail::value_t<InputBufferIt>>::type;
+    using BlevBufferDstOutT =
+      typename ::cuda::std::conditional<memcpy, void *, cub::detail::value_t<OutputBufferIt>>::type;
```
Here, I changed `void *` to `uint8_t *` because a template check was giving a compile error on using void as a type in another templated type resolution conditional expression even though was `memcpy=true` and the other branch was not taken.

elstehle replied:
I think we should fix the root cause of that issue. I left a suggestion [here](https://github.com/NVIDIA/cub/pull/675#discussion_r1170015035). Same goes for the type alias in the `agent`, as otherwise it could fail to compile if users provided pointer-to-void for a buffer's source.

mfbalin commented on cub/agent/agent_batch_memcpy.cuh:
```diff
@@ -287,6 +289,82 @@ VectorizedCopy(int32_t thread_rank, void *dest, ByteOffsetT num_bytes, const voi
 }
 
+template <bool memcpy,
+          uint32_t BLOCK_THREADS,
```
I changed it to LOGICAL_WARP_SIZE to match the vectorizedload function template parameter.

mfbalin commented:
When I use a fancy output iterator, the test doesn't do anything at all, do you have any idea why that may be?

mfbalin commented on cub/agent/agent_batch_memcpy.cuh:
```diff
@@ -287,6 +289,82 @@ VectorizedCopy(int32_t thread_rank, void *dest, ByteOffsetT num_bytes, const voi
   }
 }
 
+template <bool memcpy,
```
Changed to IsMemcpy.

elstehle commented on cub/device/dispatch/dispatch_batch_memcpy.cuh:
```diff
@@ -100,10 +103,13 @@ __launch_bounds__(int(ChainedPolicyT::ActivePolicy::AgentLargeBufferPolicyT::BLO
   using ActivePolicyT = typename ChainedPolicyT::ActivePolicy::AgentLargeBufferPolicyT;
   using BufferSizeT   = cub::detail::value_t<BufferSizeIteratorT>;
+  /// Internal load/store type. For byte-wise memcpy, a single-byte type
+  using AliasT = typename ::cuda::std::
+    conditional<IsMemcpy, char, cub::detail::value_t<cub::detail::value_t<InputBufferIt>>>::type;
```
We need to avoid the template instantiation of `iterator_traits` (via `cub::detail::value_t`) by deferring the `iterator_traits::type` member type access to avoid the compilation error for iterator-to-pointer-to-void types (`error: reference to void is not allowed`).
```suggestion
using AliasT = typename ::cuda::std::conditional<
    IsMemcpy,
    std::iterator_traits<char *>,
    std::iterator_traits<cub::detail::value_t<InputBufferIt>>>::type::value_type;
```

senior-zero reviewed (changes requested):
Thank you for the contribution! I've left a few minor comments below. Once those are addressed, I'll start testing.

mfbalin commented on cub/agent/agent_batch_memcpy.cuh:
```diff
@@ -287,6 +289,82 @@ VectorizedCopy(int32_t thread_rank, void *dest, ByteOffsetT num_bytes, const voi
+          uint32_t LOGICAL_WARP_SIZE,
+          typename InputBufferT,
+          typename OutputBufferT,
+          typename OffsetT,
```
Even after changing all instances of `a[i]` into `*(a + i)`, the code still doesn't compile with this test, all instances of `copy_items` and `read_item` and `write_item` fail with the error below:

`thrust/dependencies/cub/cub/cmake/../../cub/agent/agent_batch_memcpy.cuh(320): error: no operator "+=" matches these operands
            operand types are: thrust::device_reference<thrust::detail::normal_iterator<thrust::device_ptr<object_with_non_trivial_ctor>>> += char
          detected during:
            instantiation of "void cub::CUB_200200_800_NS::detail::copy_items<IsMemcpy,LOGICAL_WARP_SIZE,OffsetT,InputBufferT,OutputBufferT,<unnamed>>(InputBufferT, OutputBufferT, OffsetT, OffsetT) [with IsMemcpy=false, LOGICAL_WARP_SIZE=32U, OffsetT=char, InputBufferT=thrust::device_reference<thrust::detail::normal_iterator<thrust::device_ptr<object_with_non_trivial_ctor>>>, OutputBufferT=thrust::device_reference<thrust::detail::normal_iterator<thrust::device_ptr<object_with_non_trivial_ctor>>>, <unnamed>=0]" `

mfbalin commented:
> Thank you for the contribution! I've left a few minor comments below. Once those are addressed, I'll start testing.

I must have addressed all your reviews with the last commit. Thanks for the review, this is a learning opportunity to write production-quality code.

senior-zero reviewed (approved):
Thank you for addressing the review notes! We can start testing now. Please, make sure to squash your commits before merging.

elstehle commented:
Thanks a lot for your contribution and addressing all the review comments, @mfbalin! We're now [testing your PR](https://github.com/NVIDIA/thrust/pull/1925).
